amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+            shares: u128,
+            min_amounts: Vec<u128>,
+            to: AccountId,
+        ) -> Result<Vec<u128>, StablePoolError> {
+            let amounts = math::compute_withdraw_amounts_for_lp(
+                shares,
+                &self.reserves(),
+                self.psp22.total_supply(),
+            )?;
+
+            // Check if enough tokens are withdrawn
+            ensure!(
+                amounts
+                    .iter()
+                    .zip(min_amounts.iter())
+                    .all(|(amount, min_amount)| amount >= min_amount),
+                StablePoolError::InsufficientOutputAmount
+            );
+
+            // transfer tokens
+            for (&token, &amount) in self.pool.tokens.iter().zip(amounts.iter()) {
+                self.token_by_address(token).transfer(to, amount, vec![])?;
+            }
+
+            // update reserves
+            for (i, &amount) in amounts.iter().enumerate() {
+                self.decrease_reserve(i, amount)?;
+            }
+
+            // Burn liquidity
+            let events = self.psp22.burn(self.env().caller(), shares)?;
+            self.emit_events(events);
+
+            Ok(amounts)
+        }
+
+        #[ink(message)]
+        fn remove_liquidity_by_amounts(
+            &mut self,
+            max_share_amount: u128,
+            amounts: Vec<u128>,
+            to: AccountId,
+        ) -> Result<(u128, u128), StablePoolError> {
+            // Make sure rates are up to date before we attempt any calculations
+            self.update_rates();
+            if amounts.len() != self.pool.tokens.len() {
+                return Err(StablePoolError::IncorrectAmountsCount);
+            }
+
+            let rates = self.get_scaled_rates()?;
+
+            // calc comparable amounts
+            let (shares_to_burn, fee_part) = math::rated_compute_lp_amount_for_withdraw(
+                &rates,
+                &amounts,
+                &self.reserves(),
+                self.psp22.total_supply(),
+                Some(&self.pool.fees),
+                self.amp_coef(),
+            )?;
+            // check max shares
+            if shares_to_burn > max_share_amount {
+                return Err(StablePoolError::InsufficientLiquidityBurned);
+            }
+            // burn shares
+            let events = self.psp22.burn(self.env().caller(), shares_to_burn)?;
+            self.emit_events(events);
+            // mint admin fee
+            if let Some(fee_to) = self.fee_to() {
+                let admin_fee = self.pool.fees.admin_trade_fee(fee_part)?;
+                if admin_fee > 0 {
+                    let events = self.psp22.mint(fee_to, admin_fee)?;
+                    self.emit_events(events);
+                }
+            }
+            // transfer tokens
+            for (&token, &amount) in self.pool.tokens.iter().zip(amounts.iter()) {
+                self.token_by_address(token).transfer(to, amount, vec![])?;
+            }
```

Comment:
It shouldn't make a difference here (b/c it's non-reentrant) but it's a good(security) practice to transfer tokens at the end. i.e. decrease reserves first and transfer later.